crates/polars-core/src/chunked_array/logical/categorical/string_cache.rs

```diff
@@ -11,80 +11,101 @@ use crate::datatypes::PlIdHashMap;
 use crate::hashing::_HASHMAP_INIT_SIZE;
 use crate::prelude::InitHashMaps;
 
-/// We use atomic reference counting
-/// to determine how many threads use the string cache
-/// if the refcount is zero, we may clear the string cache.
-pub(crate) static USE_STRING_CACHE: AtomicU32 = AtomicU32::new(0);
+/// We use atomic reference counting to determine how many threads use the
+/// string cache. If the refcount is zero, we may clear the string cache.
+static STRING_CACHE_REFCOUNT: AtomicU32 = AtomicU32::new(0);
+static STRING_CACHE_ENABLED: AtomicBool = AtomicBool::new(false);
 static STRING_CACHE_UUID_CTR: AtomicU32 = AtomicU32::new(0);
 
-/// RAII for the string cache
-/// If an operation creates categoricals and uses them in a join
-/// or comparison that operation must hold this cache via
-/// `let handle = IUseStringCache::hold()`
-/// The cache is valid until `handle` is dropped.
+/// Enable the global string cache as long as the object is alive ([RAII]).
+///
+/// # Examples
+///
+/// Enable the string cache by initializing the object:
+///
+/// ```
+/// use polars_core::StringCacheHolder;
+///
+/// let _sc = StringCacheHolder::hold();
+/// ```
+///
+/// The string cache is enabled until `handle` is dropped.
 ///
 /// # De-allocation
+///
 /// Multiple threads can hold the string cache at the same time.
-/// The contents of the cache will only get dropped when no
-/// thread holds it.
-pub struct IUseStringCache {
+/// The contents of the cache will only get dropped when no thread holds it.
+///
+/// [RAII]: https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization
+pub struct StringCacheHolder {
     // only added so that it will never be constructed directly
     #[allow(dead_code)]
     private_zst: (),
 }
 
-impl Default for IUseStringCache {
+impl Default for StringCacheHolder {
     fn default() -> Self {
         Self::hold()
     }
 }
 
-impl IUseStringCache {
+impl StringCacheHolder {
     /// Hold the StringCache
-    pub fn hold() -> IUseStringCache {
-        enable_string_cache(true);
-        IUseStringCache { private_zst: () }
+    pub fn hold() -> StringCacheHolder {
+        increment_string_cache_refcount();
+        StringCacheHolder { private_zst: () }
     }
 }
 
-impl Drop for IUseStringCache {
+impl Drop for StringCacheHolder {
     fn drop(&mut self) {
-        enable_string_cache(false)
+        decrement_string_cache_refcount();
     }
 }
 
-pub fn with_string_cache<F: FnOnce() -> T, T>(func: F) -> T {
-    enable_string_cache(true);
-    let out = func();
-    enable_string_cache(false);
-    out
+fn increment_string_cache_refcount() {
+    STRING_CACHE_REFCOUNT.fetch_add(1, Ordering::Release);
+}
+fn decrement_string_cache_refcount() {
+    let previous = STRING_CACHE_REFCOUNT.fetch_sub(1, Ordering::Release);
+    if previous == 0 || previous == 1 {
+        STRING_CACHE_REFCOUNT.store(0, Ordering::Release);
+        STRING_CACHE.clear()
```

Comment:
Pretty sure this is race condition. Someone could have already incremented the `STRING_CACHE_REFCOUNT` again and used it in the meantime, and now we clear it.

I think we should just wrap the `STRING_CACHE_REFCOUNT` in a mutex, and then hold the mutex while clearing the cache.

It's not like we increment/decrement the string cache refcount so often that an atomic is really necessary here.

Reply:
Yeah this code seemed questionable to me too (it was already there). I can take another look and refactor this in one go, thanks for the suggestion!

Reply:
Yep.